packages/edgeworth/src/analysis/SubstanceAnalysis.ts

```diff
@@ -599,6 +600,31 @@ export const dedupStmts = (prog: SubProg<A>): SubProg<A> => ({
   ),
 });
 
+/**
+ * Remove duplicated SynthesizedSubstances.
+ * @param progs a list of SynthesizedSubstance programs
+ * @returns a list of SynthesizedSubstance programs without duplicates
+ */
+export const dedupSynthesizedSubstances = (
+  progs: SynthesizedSubstance[]
+): SynthesizedSubstance[] => {
+  // Find duplicated programs by comparing pretty-printed statements
+  const stmtsComparator = (
+    p1: SynthesizedSubstance,
+    p2: SynthesizedSubstance
+  ) => {
+    return (
+      p1.prog.statements.reduce<string>(
+        (acc, val) => acc + prettyStmt(val),
+        ""
+      ) ===
+      p2.prog.statements.reduce<string>((acc, val) => acc + prettyStmt(val), "")
```

Comment:
```suggestion
    return prettySubstance(p1.prog) === prettySubstance(p2.prog)
```

`prettySubstance` is available for pretty-printing a Substance program. I'd suggest using it to be more concise.